File: android/src/main/java/com/tozny/plugin_tozny/E3dbSerializer.java
package com.tozny.plugin_tozny;

import android.os.Build;

import androidx.annotation.RequiresApi;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.HashMap;
import java.util.Map;
import com.tozny.e3db.*;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.json.JSONObject;

public class E3dbSerializer {
    private static final ObjectMapper mapper = new ObjectMapper();
    private static DateFormat dateFormatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZZZZZ");

    static String recordMetaToJson(RecordMeta meta) {
        Map<String, Object> metaMap = new HashMap<>();
        metaMap.put("record_id", meta.recordId().toString());
        metaMap.put("user_id", meta.userId().toString());
        metaMap.put("writer_id", meta.writerId().toString());
        metaMap.put("created", dateFormatter.format(meta.created()));
        metaMap.put("last_modified", dateFormatter.format(meta.lastModified()));
        metaMap.put("version", meta.version());
        metaMap.put("type", meta.type());
        metaMap.put("plain", new JSONObject(meta.plain()).toString());
        metaMap.put("file_meta", fileMetaToJson(meta.file()));
        try {
            return mapper.writeValueAsString(metaMap);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static String fileMetaToJson(FileMeta meta) {
        if (meta == null) {
            return null;
        }
        Map<String, Object> map = new HashMap<>();
        map.put("file_url", meta.fileUrl() != null ? meta.fileUrl() : "");
        map.put("file_name", meta.fileName());
        map.put("checksum", meta.checksum());
        map.put("compression", meta.compression());
        map.put("long", meta.size());
        try {
            return mapper.writeValueAsString(map);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }


    static String recordToJson(Record record) {
        Map<String, Object> map = new HashMap<>();
        map.put("meta_data", E3dbSerializer.recordMetaToJson(record.meta()));
        map.put("data", record.data());
        try {
            return mapper.writeValueAsString(map);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static Realm realmFromJson(String json) {
        try {
            ObjectMapper mapper = new ObjectMapper();
            RealmConfigSerializer realmConfig = mapper.readValue(json, RealmConfigSerializer.class);
            return new Realm(realmConfig.realmName, realmConfig.appName, realmConfig.brokerTargetURL, realmConfig.apiURL);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static String partialIdClientToJson(PartialIdentityClient id) {
        Map<String, Object> map = new HashMap<>();
        try {
            map.put("client_credentials", id.getClient().getConfig().json());
            map.put("identity_config", E3dbSerializer.idConfigToJson(id.getIdentityConfig()));
            return mapper.writeValueAsString(map);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static String idClientToJson(IdentityClient id) {
        Map<String, Object> map = new HashMap<>();
        try {
            map.put("client_credentials", id.getClient().getConfig().json());
            map.put("identity_config", E3dbSerializer.idConfigToJson(id.getIdentityConfig()));
            map.put("user_agent_token", E3dbSerializer.userAgentTokenToJson(id.getToken()));
            return mapper.writeValueAsString(map);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static String userAgentTokenToJson(AgentToken tok) {
        Map<String, Object> map = new HashMap<>();
        try {
            map.put("access_token", tok.getToken());
            map.put("token_type", tok.getTokenType());
            map.put("expiry", tok.getExpiry());
            return mapper.writeValueAsString(map);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static String idConfigToJson(IdentityConfig id) {
        try {
            Map<String, Object> map = new HashMap<>();
            map.put("api_url", id.getApiURL().getPath());
            map.put("app_name", id.getAppName());
            map.put("broker_target_url", id.getBrokerTargetUrl());
            map.put("realm_name", id.getRealmName());
            map.put("user_id", id.getUserId());
            map.put("username", id.getUsername());

            return mapper.writeValueAsString(map);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: ios/Classes/E3dbSerializer.swift
import E3db

public class E3dbSerializer {
    // 
    static func realmFromJson(json: String) -> Application {
        let jsonData = json.data(using: .utf8)!
        let realmConfig: RealmConfigSerializer = try! JSONDecoder().decode(RealmConfigSerializer.self, from: jsonData)
        return Application(apiUrl: realmConfig.apiURL, appName: realmConfig.appName, realmName: realmConfig.realmName, brokerTargetUrl: realmConfig.brokerTargetURL)
    }
    
    static func idClientToJson(id: Identity) -> String {
        var data: [String: Any] = [:]
        data.updateValue(E3dbSerializer.configToJson(storageConfig: FlutterConfig.decodeToFlutterConfig(e3dbConfig: id.idConfig.storageConfig)), forKey: "client_credentials")
        data.updateValue(E3dbSerializer.idConfigToJson(idConfig: id.idConfig), forKey: "identity_config")
        data.updateValue(E3dbSerializer.userAgentTokenToJson(tok: id.agentInfo()), forKey: "user_agent_token")
        let jsonData = try! JSONSerialization.data(withJSONObject: data, options: .prettyPrinted)
        let jsonString = String(data: jsonData, encoding: .utf8)!
        return jsonString
    }
    
    static func partialIdClientToJson(id: PartialIdentity) -> String {
        var data: [String: Any] = [:]
        data.updateValue(E3dbSerializer.configToJson(storageConfig: FlutterConfig.decodeToFlutterConfig(e3dbConfig: id.idConfig.storageConfig)), forKey: "client_credentials")
        data.updateValue(E3dbSerializer.idConfigToJson(idConfig: id.idConfig), forKey: "identity_config")
        let jsonData = try! JSONSerialization.data(withJSONObject: data, options: .prettyPrinted)
        let jsonString = String(data: jsonData, encoding: .utf8)!
        return jsonString
    }

    static func configToJson(storageConfig: FlutterConfig) -> String {
            let jsonData = try! JSONEncoder().encode(storageConfig)
            let jsonString = String(data: jsonData, encoding: .utf8)!
            return jsonString
    }
    
    static func recordMetaToJson(meta: Meta) -> String {
        var data: [String: Any] = [:]

        data.updateValue(meta.recordId.description, forKey: "record_id")
        data.updateValue(meta.userId.description, forKey: "user_id")
        data.updateValue(meta.writerId.description, forKey: "writer_id")
        data.updateValue(meta.created.description, forKey: "created")
        data.updateValue(meta.lastModified.description, forKey: "last_modified")
        data.updateValue(meta.version, forKey: "version")
        data.updateValue(meta.type, forKey: "type")

        let plainJsonData = try! JSONSerialization.data(withJSONObject: meta.plain!, options: .prettyPrinted)
        let plainJsonString = String(data: plainJsonData, encoding: .utf8)!

        data.updateValue(plainJsonString, forKey: "plain")
        data.updateValue(fileMetaToJson(meta: meta.fileMeta), forKey: "file_meta")
    
        let jsonData = try! JSONSerialization.data(withJSONObject: data, options: .prettyPrinted)
        let jsonString = String(data: jsonData, encoding: .utf8)!
        return jsonString
    }
    
    static func recordToJson(record: Record) -> String {
        var data: [String: Any] = [:]
        data.updateValue(E3dbSerializer.recordMetaToJson(meta: record.meta), forKey: "meta_data")
        data.updateValue(record.data, forKey: "data")
        let jsonData = try! JSONSerialization.data(withJSONObject: data, options: .prettyPrinted)
        let jsonString = String(data: jsonData, encoding: .utf8)!
        return jsonString
    }

    static func fileMetaToJson(meta: FileMeta?) -> String? {
        if meta == nil {
            return nil
        }
        var data: [String: Any] = [:]
        
        /// If meta is not `nil`, then assume we can force unwrap its values
        data.updateValue(meta!.fileUrl?.description, forKey: "file_url")
        data.updateValue(meta!.fileName!, forKey: "fileName")
        data.updateValue(meta!.checksum, forKey: "checksum")
        data.updateValue(meta!.compression, forKey: "compression")
        data.updateValue(meta!.size?.description, forKey: "long")
        let jsonData = try! JSONSerialization.data(withJSONObject: data, options: .prettyPrinted)
        let jsonString = String(data: jsonData, encoding: .utf8)!
        return jsonString
    }
    
    static func idConfigToJson(idConfig: IdentityConfig) -> String {
        let jsonData = try! JSONEncoder().encode(idConfig)
        let jsonString = String(data: jsonData, encoding: .utf8)!
        return jsonString
    }
    
    static func userAgentTokenToJson(tok: AgentToken) -> String {
        let jsonData = try! JSONEncoder().encode(tok)
        var intermediateFlutterAgentToken = try! JSONDecoder().decode(IntermediateFlutterAgentToken.self, from: jsonData)
        
        /// Convert Swift `timeIntervalSinceReferenceDate` to `timeIntervalSince1970` in milliseconds
        /// Apple considers its 0 time as 1 January 2001 instead of 1 January 1971, this performs the conversion from the
        /// former to the latter.
        intermediateFlutterAgentToken.expiryAsUnixEpoch = Int(intermediateFlutterAgentToken.expiry.timeIntervalSince1970*1000)
        let flutterAgentTokenMap = IntermediateFlutterAgentToken.encodeWithoutDateType(flutterAgentToken: intermediateFlutterAgentToken)
        
        /// Need to respect Swift's Codable protocol and parity between Android and iOS plugins, so another level of indirection
        /// is performed in order to remove the `FlutterAgentToken.expiryAsUnixEpoch` field and make the `expiry` field
        /// `Int` instead of a `String`
        let flutterAgentToken = FlutterAgentToken(accessToken: flutterAgentTokenMap["access_token"]!, tokenType: flutterAgentTokenMap["token_type"]!, expiry: Int(flutterAgentTokenMap["expiry"]!)!)
                
        let flutterAgentTokenData = try! JSONEncoder().encode(flutterAgentToken)

        let jsonString = String(data: flutterAgentTokenData, encoding: .utf8)!
        return jsonString
    }
}
